### PR TITLE
PR: xde does not update VPC mappings when a Port is deleted (#294)

### PR DESCRIPTION
**Pull request by rzezeski**

### CODE REVIEW

**Comment by rzezeski** [comment]
I tested this by standing up a guest, verifying its VPC mapping is there and that it can send traffic, then I stopped the guest and verified the port and VPC mapping are gone.

* Start guest.
```
rpz@kalm:~/oxidecomputer/opte/xde$ pfexec opteadm list-ports                             
LINK                             MAC ADDRESS              IPv4 ADDRESS     EXTERNAL IPv4    IPv6 ADDRESS                             EXTERNAL IPv6                            ST
ATE                                                                                     
opte3                            A8:40:25:F1:6A:F2        172.30.0.5       10.0.0.227       None                                     None                                     running                              

rpz@kalm:~/oxidecomputer/opte/xde$ pfexec opteadm dump-v2p
Virtual to Physical Mappings
======================================================================

VPC 15761424
----------------------------------------------------------------------

IPv4 mappings
----------------------------------------------------------------------
VPC IP                   VPC MAC ADDR      UNDERLAY IP
172.30.0.5               A8:40:25:F1:6A:F2 fd00:1122:3344:101::1

IPv6 mappings
----------------------------------------------------------------------
VPC IP                   VPC MAC ADDR      UNDERLAY IP

rpz@kalm:~/oxidecomputer/opte/xde$ pfexec opteadm dump-tcp-flows -p opte3                
FLOW                                             STATE        HITS     SEGS IN  SEGS OUT BYTES IN   BYTES OUT 
TCP:172.30.0.5:22:10.0.0.10:53242                ESTABLISHED  124      69       56       7731       7961      
TCP:172.30.0.5:46946:52.10.128.69:80             TIME_WAIT    10       5        6        833        479    

rpz@kalm:~/oxidecomputer/opte/xde$ pfexec opteadm dump-uft -p opte3                      
UFT Inbound: 7/8096
----------------------------------------------------------------------
PROTO  SRC IP           SPORT  DST IP           DPORT  HITS     XFORMS                
ICMP   10.0.0.1         0      10.0.0.227       0      2        hdr: NAT, body:       
TCP    10.0.0.10        52303  10.0.0.227       22     0        hdr: NAT, body:       
TCP    10.0.0.10        53242  10.0.0.227       22     68       hdr: NAT, body:       
TCP    52.10.128.69     80     10.0.0.227       46946  4        hdr: NAT, body:       
TCP    52.10.128.69     443    10.0.0.227       50614  14       hdr: NAT, body:       
UDP    8.8.8.8          53     10.0.0.227       40358  1        hdr: NAT, body:       
UDP    8.8.8.8          53     10.0.0.227       60974  1        hdr: NAT, body:       

UFT Outbound: 4/8096
----------------------------------------------------------------------
PROTO  SRC IP           SPORT  DST IP           DPORT  HITS     XFORMS                
TCP    172.30.0.5       22     10.0.0.10        53242  55       hdr: NAT, body:       
TCP    172.30.0.5       46946  52.10.128.69     80     5        hdr: NAT, body:       
UDP    172.30.0.5       40358  8.8.8.8          53     1        hdr: NAT, body:       
UDP    172.30.0.5       60974  8.8.8.8          53     1        hdr: NAT, body:  
```


* Stop guest.
```
rpz@kalm:~/oxidecomputer/opte/xde$ pfexec opteadm list-ports
LINK                             MAC ADDRESS              IPv4 ADDRESS     EXTERNAL IPv4    IPv6 ADDRESS                             EXTERNAL IPv6                            ST
ATE   

rpz@kalm:~/oxidecomputer/opte/xde$ pfexec opteadm dump-v2p                                                                                                                      
Virtual to Physical Mappings
======================================================================                                                                                                          
                                            
VPC 15761424                       
----------------------------------------------------------------------
                                                                                        
IPv4 mappings                                                                                                                                                                   
----------------------------------------------------------------------
VPC IP                   VPC MAC ADDR      UNDERLAY IP
                                                                                                                                                                                
IPv6 mappings               
----------------------------------------------------------------------
VPC IP                   VPC MAC ADDR      UNDERLAY IP
```

File: oxide-vpc/src/api.rs [hunk 1]
```diff
@@ -276,13 +276,22 @@ pub struct PhysNet {
     pub vni: Vni,
 }
 
-/// The physical address for a guest.
+/// The physical address for a guest, minus the VNI.
+///
+/// We save space in the VPC mappings by grouping guest
+/// Virtual-to-Physical mappings by VNI.
```
Comment (bnaecker) [inline oxide-vpc/src/api.rs:282]: These are really helpful, I never quite understood why we split these up this way.

File: oxide-vpc/src/engine/overlay.rs [hunk 2]
```diff
@@ -428,6 +430,25 @@ impl VpcMappings {
         }
     }
 
+    /// Delete the mapping for the given VIP in the given VNI.
+    ///
+    /// Return the existing entry, if there is one.
+    pub fn del(&self, vip: &IpAddr, phys: PhysNet) -> Option<PhysNet> {
```
Comment (bnaecker) [inline oxide-vpc/src/engine/overlay.rs:436]: I believe this type is `Copy`, but me might want to take this by reference. The method doesn't really need to take ownership of that, and would be nice to signal in the interface.

**Review by bnaecker — COMMENTED** [review]
Nice, thanks! One nit about by-val / by-ref, but otherwise LGTM.

File: oxide-vpc/src/engine/overlay.rs (hunk 2 above)
Comment (rzezeski) [reply oxide-vpc/src/engine/overlay.rs]: Yep, meant to do that, thanks for catching.

**Review by bnaecker — APPROVED** [review]

**Pull request merged**